vllm/model_executor/layers/quantization/gptq.py

```diff
@@ -0,0 +1,226 @@
+                "input_dim": 0,
+                "output_dim": 1,
+                "packed_dim": 0,
+                "pack_factor": self.quant_config.pack_factor,
+            })
+        g_idx = Parameter(
+            torch.tensor(
+                [i // self.quant_config.group_size for i in range(input_size)],
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(g_idx, {"input_dim": 0})
+        tp_size = get_tensor_model_parallel_world_size()
+        if parallel_type == "row" and tp_size > 1 and (self.quant_config.desc_act
+                and self.quant_config.group_size != -1):
+            input_size = input_size * tp_size
+            use_exllama = Parameter(torch.tensor(False, dtype=torch.bool, device="cuda"), requires_grad=False)
+        else:
+            use_exllama = Parameter(torch.tensor(True, dtype=torch.bool, device="cuda"), requires_grad=False)
+        if self.quant_config.desc_act or self.quant_config.group_size == -1:
+            input_dim = None
+        else:
+            input_dim = 0
+        group_size = self.quant_config.group_size if self.quant_config.group_size != -1 else input_size
+        qzeros = Parameter(
+            torch.empty(
+                input_size // group_size,
+                output_size // self.quant_config.pack_factor,
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(
+            qzeros, {
+                "input_dim": input_dim,
+                "output_dim": 1,
+                "packed_dim": 1,
+                "pack_factor": self.quant_config.pack_factor,
+            })
+        scales = Parameter(
+            torch.empty(
+                input_size // group_size,
+                output_size,
+                device="cuda",
+                dtype=params_dtype,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(scales, {
+            "input_dim": input_dim,
+            "output_dim": 1,
+        })
+        return {
+            "qweight": qweight,
+            "g_idx": g_idx,
+            "qzeros": qzeros,
+            "scales": scales,
+            "use_exllama": use_exllama,
+        }
+
+    def apply_weights(self,
+                      weights: Dict[str, torch.Tensor],
+                      x: torch.Tensor,
+                      bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        qweight = weights["qweight"]
+        height, width = weights["qweight"].shape
+        out_shape = x.shape[:-1] + (qweight.shape[-1], )
+        reshaped_x = x.reshape(-1, x.shape[-1])
+        if weights["use_exllama"]:
+            if "q4" not in weights:
+                if not self.quant_config.desc_act:
+                    none_tensor = torch.empty((1, 1), device="meta")
+                    weights["q4"] = ops.make_q_matrix(
```

Comment:
Can we create the weights in `create_weights`? We can change the API of `create_weights` to also return some non-parameter tensors to achieve this.

Reply:
This function does some weight reordering and thus has to be called after weight is loaded. We can remove the `q4` as commented above, but still we need to perform the reorder here.